lxd/instance/drivers/driver_qemu.go

```diff
@@ -2589,6 +2589,158 @@ func (d *qemu) nvramPath() string {
+	}
+
+	// If "capsh" tool is present in the system, let's drop
+	// all capabilities except "cap_dac_override".
+	var cmdName string
+	var cmdArgs []string
+	if _, err := exec.LookPath("capsh"); err == nil {
+		cmdName = "capsh"
+		cmdArgs = []string{
+			"--drop=cap_chown,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linux_immutable,cap_net_bind_service,cap_net_broadcast,cap_net_admin,cap_net_raw,cap_ipc_lock,cap_ipc_owner,cap_sys_module,cap_sys_rawio,cap_sys_chroot,cap_sys_ptrace,cap_sys_pacct,cap_sys_admin,cap_sys_boot,cap_sys_nice,cap_sys_resource,cap_sys_time,cap_sys_tty_config,cap_mknod,cap_lease,cap_audit_write,cap_audit_control,cap_setfcap,cap_mac_override,cap_mac_admin,cap_syslog,cap_wake_alarm,cap_block_suspend,cap_audit_read,cap_perfmon,cap_bpf,cap_checkpoint_restore",
+			"--", "-c", pyUEFIVarsCmd}
+	} else {
+		cmdName = "sh"
+		cmdArgs = []string{"-c", pyUEFIVarsCmd}
+	}
+
+	err := shared.RunCommandWithFds(context.TODO(), stdin, stdout, cmdName, cmdArgs...)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// EFIVarsRead does not make any sense for containers.
+func (d *qemu) EFIVarsRead() (api.InstanceEFI, error) {
+	efiVarsPath := d.nvramPath()
+
+	var stdout bytes.Buffer
+	err := callPyUEFIVars(nil, &stdout, efiVarsPath)
+	if err != nil {
+		return api.InstanceEFI{}, err
+	}
+
+	pyUEFIVars := PyUEFIVars{}
+	err = json.Unmarshal(stdout.Bytes(), &pyUEFIVars)
+	if err != nil {
+		return api.InstanceEFI{}, err
+	}
+
+	if pyUEFIVars.Version != 2 {
+		return api.InstanceEFI{}, fmt.Errorf("python-uefivars utility version is not compatible with LXD server")
+	}
+
+	instanceEFI := api.InstanceEFI{}
+	instanceEFI.Variables = make(map[string]*api.InstanceEFIVariable)
+
+	for _, v := range pyUEFIVars.Variables {
+		key := v.Name + "-" + v.GUID
+		instanceEFI.Variables[key] = &api.InstanceEFIVariable{
+			Data:      v.Data,
+			Attr:      v.Attr,
+			Timestamp: v.Timestamp,
+			Digest:    v.Digest,
+		}
+	}
+
+	return instanceEFI, nil
+}
+
+// EFIVarsUpdate does not make any sense for containers.
+func (d *qemu) EFIVarsUpdate(newEFIVarsSet api.InstanceEFI) error {
+	if d.IsRunning() {
+		return fmt.Errorf("EFI variables editing is allowed for stopped VM instances only")
+	}
+
+	if !d.architectureSupportsUEFI(d.architecture) {
+		return fmt.Errorf("EFI is not supported for this instance architecture")
+	}
+
+	if shared.IsTrue(d.expandedConfig["security.csm"]) {
+		return fmt.Errorf("EFI is disabled when CSM mode is active")
+	}
+
+	efiVarsPath := d.nvramPath()
+
+	var stdout bytes.Buffer
+	err := callPyUEFIVars(nil, &stdout, efiVarsPath)
+	if err != nil {
+		return err
+	}
+
+	pyUEFIVars := PyUEFIVars{}
+	err = json.Unmarshal(stdout.Bytes(), &pyUEFIVars)
+	if err != nil {
+		return err
+	}
+
+	pyUEFIVars = PyUEFIVars{
+		Version:   2,
+		Variables: make([]PyUEFIVar, 0),
+	}
+
+	// see also (e InstanceEFI) CheckValid()
```

Comment:
im not understanding this comment?

Reply:
Comment was removed during code reorganization ;-)

Idea was that in the `CheckValid()` method I did all the validation checks and wanted to pay attention to it.